### ISSUE TITLE
Allow backtracking when using generics

### ISSUE THREAD

**Issue by marovira** [astral-sh/ty#3119]
### Summary

I tried looking through the issues list, but couldn't find anything similar. 
Consider the following code:
```python
import typing

T = typing.TypeVar("T")

def convert_to_list(var: T | list[T] | tuple[T, ...]) -> list[T]:
    if isinstance(var, list):
        return var
    if isinstance(var, tuple):
        return list(var) # ty: ignore[invalid-argument-type]
    return [var]

def foo(x: float | list[float] | tuple[float, ...] | str | list[str] | tuple[str, ...]) -> None:
    typing.reveal_type(x)
    as_list = convert_to_list(x)
    typing.reveal_type(as_list)
    for elem in as_list:
        print(elem)
```

If I run `ty check` on this, I get the following output:
```
info[revealed-type]: Revealed type
  --> main.py:13:24
   |
12 | def foo(x: float | list[float] | tuple[float, ...] | str | list[str] | tuple[str, ...]) -> None:
13 |     typing.reveal_type(x)
   |                        ^ `int | float | list[int | float] | tuple[int | float, ...] | str | list[str] | tuple[str, ...]`
14 |     as_list = convert_to_list(x)
15 |     typing.reveal_type(as_list)
   |

info[revealed-type]: Revealed type
  --> main.py:15:24
   |
13 |     typing.reveal_type(x)
14 |     as_list = convert_to_list(x)
15 |     typing.reveal_type(as_list)
   |                        ^^^^^^^ `list[Unknown]`
16 |     for elem in as_list:
17 |         print(elem)
   |

Found 2 diagnostics
```

The first `reveal_type` is correct, but the second `reveal_type` is wrong. Ty is deducing the type to be `list[Unknown]` when in reality it should be `list[float] | list[str]`. If I add the type annotation to `as_list`, the type is still revealed as `list[Unknown]`. You can see this behaviour on [this playground](https://play.ty.dev/17ccdd6d-1b85-4bf1-a51b-a9d59fe2b0f6)

> [!NOTE]
> Please ignore the `# ty: ignore` comment. That is to silence an error from ty that is already being tracked in a separate issue.

### Version

0.0.24 and 146340ca2

**Comment by carljm** [comment]
Thanks! I think this is basically #1703.
